### PR TITLE
Fix loss normalization for padded actions in ACT, Diffusion, and MultiTaskDiT

### LINKED ISSUE #3353: Incorrect loss normalization when masking padded actions in ACT (mean over padded horizon)
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell

```

### Description

The current loss computation in ACT applies a mask to padded actions but still performs a global .mean() reduction. This results in incorrect normalization when sequences contain padding.

Specifically, padded action entries are zeroed out before reduction, but they are still included in the denominator when computing the mean. This causes the loss to be underestimated for sequences with more padding i.e., 60 to 70% padding.

### Context & Reproduction

**Current implementation (line 114 of modelling_act)**
https://github.com/huggingface/lerobot/blob/main/src/lerobot/policies/act/modeling_act.py#L144
`l1_loss = (
    F.l1_loss(batch[ACTION], actions_hat, reduction="none")
    * ~batch["action_is_pad"].unsqueeze(-1)
).mean()`

**Issue**:

- action_is_pad correctly masks padded timesteps
- however, .mean() still divides by the total number of elements (including padded ones)
- this effectively dilutes the contribution of valid (non-padded) actions

**Expected behavior**

The loss should be normalized only over valid (non-padded) action entries, not over the full padded horizon.

A correct formulation would be:

```
abs_err = F.l1_loss(batch[ACTION], actions_hat, reduction="none")

valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
num_valid = valid_mask.sum() * abs_err.shape[-1]

l1_loss = (abs_err * valid_mask).sum() / num_valid.clamp_min(1)
```

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

_No response_

### PR DESCRIPTION
**Pull request by he-yufeng**
## What this fixes

When `action_is_pad` masks out padded timesteps, the subsequent `.mean()` still divides by the total element count (including zeroed-out padding). With 60-70% padding this underestimates the loss by 2-3x, reducing gradient signal and hurting training convergence.

**Before (ACT example):**
```python
l1_loss = (
    F.l1_loss(batch[ACTION], actions_hat, reduction="none")
    * ~batch["action_is_pad"].unsqueeze(-1)
).mean()  # divides by ALL elements including padding
```

**After:**
```python
valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
l1_loss = (
    F.l1_loss(batch[ACTION], actions_hat, reduction="none") * valid_mask
).sum() / valid_mask.sum().clamp_min(1)  # divides by valid elements only
```

## Affected policies

| Policy | File | Fixed |
|--------|------|-------|
| ACT | `modeling_act.py:145` | Yes |
| Diffusion | `modeling_diffusion.py:376` | Yes |
| MultiTaskDiT (DDPM) | `modeling_multi_task_dit.py:689` | Yes |
| MultiTaskDiT (Flow) | `modeling_multi_task_dit.py:753` | Yes |
| TDMPC | `modeling_tdmpc.py` | Not affected (sums over time before batch mean) |

## Test plan

- [ ] `uv run pytest tests/policies -v` passes
- [ ] Verify with a batch where ~50% actions are padded: loss should be higher than before (no dilution)

Fixes #3353

### CODE REVIEW

**Review by Maximellerbach — COMMENTED** [review]
This is a good catch, 
make sure you run: `pre-commit run -a` so that the lint / coding style tests pass

other than that LGTM

File: src/lerobot/policies/act/modeling_act.py [hunk 1]
```diff
@@ -142,9 +142,8 @@ def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, dict]:
 
         actions_hat, (mu_hat, log_sigma_x2_hat) = self.model(batch)
 
-        l1_loss = (
-            F.l1_loss(batch[ACTION], actions_hat, reduction="none") * ~batch["action_is_pad"].unsqueeze(-1)
-        ).mean()
+        valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
+        l1_loss = (F.l1_loss(batch[ACTION], actions_hat, reduction="none") * valid_mask).sum() / valid_mask.sum().clamp_min(1)
```
Comment (whats2000) [inline src/lerobot/policies/act/modeling_act.py]: Running `pre-commit` only touches this one line by ruff-format. No behavior change, purely cosmetic.

```suggestion
        l1_loss = (
            F.l1_loss(batch[ACTION], actions_hat, reduction="none") * valid_mask
        ).sum() / valid_mask.sum().clamp_min(1)
```

**Review by whats2000 — COMMENTED** [review]
Running `pre-commit` only touches this one line by ruff-format. No behavior change, purely cosmetic.

**Comment by whats2000** [comment]
The denominator `valid_mask.sum()` counts timesteps (mask shape `(B, H, 1)`), while the numerator sums over the action dimension. Result: `loss_B = action_dim * loss_C` at every step. This likely includes zero-padded batches, silently rescaling L1 by `action_dim` for all existing recipes, not just padded ones.

Test results:

```text
A  pre-patch       mask-then-mean (parent commit d57c58a5)
B  patched (HEAD)  (F.l1(...) * mask).sum() / mask.sum() # commit 9cddd808
C  my proposal     divide by (mask.sum() * action_dim) to match issue #3353 and compute a per-scalar mean over valid entries.

Records: A=30 B=30 C=30 (action_dim=14)

mean loss_B / loss_C = 14.0009   stdev=0.0063
mean loss_A / loss_C = 0.9604    stdev=0.0645
mean grdn_B / grdn_C = 13.9982   stdev=0.0817
```

File: src/lerobot/policies/act/modeling_act.py [hunk 2]
```diff
@@ -142,9 +142,10 @@ def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, dict]:
 
         actions_hat, (mu_hat, log_sigma_x2_hat) = self.model(batch)
 
+        valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
         l1_loss = (
-            F.l1_loss(batch[ACTION], actions_hat, reduction="none") * ~batch["action_is_pad"].unsqueeze(-1)
-        ).mean()
+            F.l1_loss(batch[ACTION], actions_hat, reduction="none") * valid_mask
+        ).sum() / valid_mask.sum().clamp_min(1)
```
Comment (whats2000) [inline src/lerobot/policies/act/modeling_act.py:148]: The denominator counts timesteps only, but `abs_err` sums over `action_dim`. With ALOHA `action_dim=14` the patched loss is `14 × per-scalar mean`, so every recipe silently runs at `14 ×` the intended effective LR. Multiply by `abs_err.shape[-1]` to match the formula in issue #3353:

```suggestion
        abs_err = F.l1_loss(batch[ACTION], actions_hat, reduction="none")
        valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
        num_valid = valid_mask.sum() * abs_err.shape[-1]
        l1_loss = (abs_err * valid_mask).sum() / num_valid.clamp_min(1)
```

File: src/lerobot/policies/diffusion/modeling_diffusion.py [hunk 3]
```diff
@@ -374,7 +374,8 @@ def compute_loss(self, batch: dict[str, Tensor]) -> Tensor:
                     f"{self.config.do_mask_loss_for_padding=}."
                 )
             in_episode_bound = ~batch["action_is_pad"]
-            loss = loss * in_episode_bound.unsqueeze(-1)
+            mask = in_episode_bound.unsqueeze(-1)
+            return (loss * mask).sum() / mask.sum().clamp_min(1)
```
Comment (whats2000) [inline src/lerobot/policies/diffusion/modeling_diffusion.py:378]: Same issue: `mask.sum()` counts timesteps but `loss` sums over `action_dim`. Divide by `mask.sum() * loss.shape[-1]`:

```suggestion
            num_valid = mask.sum() * loss.shape[-1]
            return (loss * mask).sum() / num_valid.clamp_min(1)
```

File: src/lerobot/policies/multi_task_dit/modeling_multi_task_dit.py [hunk 4]
```diff
@@ -685,8 +685,8 @@ def compute_loss(self, model: nn.Module, batch: dict[str, Tensor], conditioning_
         loss = F.mse_loss(predicted, target, reduction="none")
 
         if self.do_mask_loss_for_padding and "action_is_pad" in batch:
-            valid_actions = ~batch["action_is_pad"]
-            loss = loss * valid_actions.unsqueeze(-1)
+            mask = ~batch["action_is_pad"].unsqueeze(-1)
+            return (loss * mask).sum() / mask.sum().clamp_min(1)
```
Comment (whats2000) [inline src/lerobot/policies/multi_task_dit/modeling_multi_task_dit.py:689]: Same issue: divide by `mask.sum() * loss.shape[-1]`:

```suggestion
            num_valid = mask.sum() * loss.shape[-1]
            return (loss * mask).sum() / num_valid.clamp_min(1)
```

File: src/lerobot/policies/multi_task_dit/modeling_multi_task_dit.py [hunk 5]
```diff
@@ -749,8 +749,8 @@ def compute_loss(self, model: nn.Module, batch: dict[str, Tensor], conditioning_
         loss = F.mse_loss(predicted_velocity, target_velocity, reduction="none")
 
         if self.do_mask_loss_for_padding and "action_is_pad" in batch:
-            valid_mask = ~batch["action_is_pad"]
-            loss = loss * valid_mask.unsqueeze(-1)
+            mask = ~batch["action_is_pad"].unsqueeze(-1)
+            return (loss * mask).sum() / mask.sum().clamp_min(1)
```
Comment (whats2000) [inline src/lerobot/policies/multi_task_dit/modeling_multi_task_dit.py:753]: Same issue: divide by `mask.sum() * loss.shape[-1]`:

```suggestion
            num_valid = mask.sum() * loss.shape[-1]
            return (loss * mask).sum() / num_valid.clamp_min(1)
```

**Review by whats2000 — CHANGES_REQUESTED** [review]
Ran this patch through `lerobot-train` on ALOHA (`action_dim=14`, `use_vae=false`, 30 steps), comparing against the pre-patch baseline and a per-scalar-mean variant.

**Issue:** the denominator `valid_mask.sum()` counts timesteps (mask shape `(B, H, 1)`) while the numerator sums over the action dimension, so the loss is silently scaled by `action_dim` at every step — including zero-padding batches.

```
mean  loss_B(this PR)        / loss_C(per-scalar mean) = 14.0009   stdev 0.0063
mean  loss_A(pre-patch mean) / loss_C                  = 0.9604    stdev 0.0645
mean  grdn_B                 / grdn_C                  = 13.9982   stdev 0.0817
```

Four inline suggestions attached for each affected site. `modeling_smolvla.py` still has the pre-patch dilution pattern and isn't covered by this PR  -> worth a separate pass.

**Review by whats2000 — CHANGES_REQUESTED** [review]
While verifying this PR I also grep'd the rest of `src/lerobot/policies/` for the same `mask-then-mean` pattern. Two more sites are affected, both out of scope for this PR but worth flagging so nothing is left behind.

## 1. `modeling_smolvla.py`

```python
if actions_is_pad is not None:
    in_episode_bound = ~actions_is_pad
    losses = losses * in_episode_bound.unsqueeze(-1)
    ...
loss = losses.mean()
```

Same issue as issue #3353 describes for ACT. Not covered by this PR. Suggested diff (same `num_valid = mask.sum() * shape[-1]` idiom as the other four sites):

```diff
         if reduction == "none":
-            # Return per-sample losses (B,) by averaging over time and action dims
-            per_sample_loss = losses.mean(dim=(1, 2))
+            # Return per-sample losses (B,) by averaging over valid (time, action) entries
+            if actions_is_pad is None:
+                per_sample_loss = losses.mean(dim=(1, 2))
+            else:
+                num_valid = ((~actions_is_pad).sum(dim=1) * losses.shape[-1]).clamp_min(1)
+                per_sample_loss = losses.sum(dim=(1, 2)) / num_valid
             loss_dict["loss"] = per_sample_loss.mean().item()
             return per_sample_loss, loss_dict
         else:
-            # Default: return scalar mean loss
-            loss = losses.mean()
+            # Default: return scalar mean loss over valid (time, action) entries
+            if actions_is_pad is None:
+                loss = losses.mean()
+            else:
+                num_valid = ((~actions_is_pad).sum() * losses.shape[-1]).clamp_min(1)
+                loss = losses.sum() / num_valid
             loss_dict["loss"] = loss.item()
             return loss, loss_dict
```

**Review by whats2000 — CHANGES_REQUESTED** [review]
Re-submitting the ACT variant-C suggestion manually.

**Comment by imstevenpmwork** [comment]
@he-yufeng Could you help us applying the remaining suggestions ? 😄

File: src/lerobot/policies/act/modeling_act.py (hunk 2 above)
Comment (whats2000) [inline src/lerobot/policies/act/modeling_act.py]:
```suggestion
        abs_err = F.l1_loss(batch[ACTION], actions_hat, reduction="none")
        valid_mask = ~batch["action_is_pad"].unsqueeze(-1)
        num_valid = valid_mask.sum() * abs_err.shape[-1]
        l1_loss = (abs_err * valid_mask).sum() / num_valid.clamp_min(1)
```

**Review by whats2000 — CHANGES_REQUESTED** [review]

**Comment by he-yufeng** [comment]
@whats2000 @imstevenpmwork Applied the ACT suggestion — it now matches the diffusion/multi_task_dit pattern (divide by timesteps × action_dim, not just timesteps). The other two were already merged in via GitHub's suggestion commits. Thanks for the careful review!

**Review by whats2000 — APPROVED** [review]
I have made a test run again looks good!

**Comment by whats2000** [comment]
> While verifying this PR I also grep'd the rest of `src/lerobot/policies/` for the same `mask-then-mean` pattern. Two more sites are affected, both out of scope for this PR but worth flagging so nothing is left behind.
> 
> ## 1. `modeling_smolvla.py`
> ```python
> if actions_is_pad is not None:
>     in_episode_bound = ~actions_is_pad
>     losses = losses * in_episode_bound.unsqueeze(-1)
>     ...
> loss = losses.mean()
> ```
> 
> Same issue as issue #3353 describes for ACT. Not covered by this PR. Suggested diff (same `num_valid = mask.sum() * shape[-1]` idiom as the other four sites):
> 
> ```diff
>          if reduction == "none":
> -            # Return per-sample losses (B,) by averaging over time and action dims
> -            per_sample_loss = losses.mean(dim=(1, 2))
> +            # Return per-sample losses (B,) by averaging over valid (time, action) entries
> +            if actions_is_pad is None:
> +                per_sample_loss = losses.mean(dim=(1, 2))
> +            else:
> +                num_valid = ((~actions_is_pad).sum(dim=1) * losses.shape[-1]).clamp_min(1)
> +                per_sample_loss = losses.sum(dim=(1, 2)) / num_valid
>              loss_dict["loss"] = per_sample_loss.mean().item()
>              return per_sample_loss, loss_dict
>          else:
> -            # Default: return scalar mean loss
> -            loss = losses.mean()
> +            # Default: return scalar mean loss over valid (time, action) entries
> +            if actions_is_pad is None:
> +                loss = losses.mean()
> +            else:
> +                num_valid = ((~actions_is_pad).sum() * losses.shape[-1]).clamp_min(1)
> +                loss = losses.sum() / num_valid
>              loss_dict["loss"] = loss.item()
>              return loss, loss_dict
> ```

I think for this issue, I will create a sub-issue with another patch PR instead with test!

**Review by imstevenpmwork — APPROVED** [review]
LGTM, thanks!

**Comment by imstevenpmwork** [comment]
Superseded by: https://github.com/huggingface/lerobot/pull/3442

**Comment by he-yufeng** [comment]
Thanks for picking this up in #3442 — glad the fix landed. Closing this one as superseded.
